### PR TITLE
99 handle language keywords

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ edition = "2018"
 quote = "1.0.21"
 # syn = "1.0.99"
 proc-macro2 = "1.0.43"
+regex = "1"
```
Comment (lastmjs) [inline Cargo.toml:13]: Put a version

File: src/abstract_canister_tree.rs [hunk 2]
```diff
@@ -46,30 +47,34 @@ impl ToTokenStream for AbstractCanisterTree {
...
+        let cross_canister_functions = self.external_canisters.to_token_streams(&self.keywords);
 
         let user_defined_code = &self.rust_code;
 
-        let heartbeat_method = self.heartbeat_method.to_token_stream();
-        let init_method = self.init_method.to_token_stream();
-        let inspect_message_method = self.inspect_message_method.to_token_stream();
-        let post_upgrade_method = self.post_upgrade_method.to_token_stream();
-        let pre_upgrade_method = self.pre_upgrade_method.to_token_stream();
+        let heartbeat_method = self.heartbeat_method.to_token_stream(context);
```
Comment (lastmjs) [inline src/abstract_canister_tree.rs:54]: Use explicit unit type

File: src/keyword.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        name.clone()
+    }
+}
+
+pub fn to_regex(keywords: &Vec<String>) -> Vec<Regex> {
+    keywords
+        .iter()
+        .map(|keyword| {
+            let keyword = regex::escape(keyword);
+            Regex::new(format!(r#"{}"#, keyword).as_str()).unwrap()
```
Comment (lastmjs) [inline src/keyword.rs:22]: This won't even work at all

File: src/keyword.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+use regex::Regex;
```
Comment (lastmjs) [inline src/keyword.rs:1]: Create an actual example with records and variants with all kinds of keywords

File: src/lib.rs [hunk 5]
```diff
@@ -51,24 +52,26 @@ pub trait ToAct {
...
+    fn to_token_streams(&self, context: C) -> Vec<TokenStream> {
+        self.iter()
+            .map(|t| t.to_token_stream(context.clone()))
+            .collect()
     }
 }
 
-impl<T: ToTokenStream> ToTokenStream for Option<T> {
-    fn to_token_stream(&self) -> TokenStream {
+impl<T: ToTokenStream<C>, C> ToTokenStream<C> for Option<T> {
```
Comment (lastmjs) [inline src/lib.rs:71]: Change C order

File: src/keyword.rs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    });
+
+    if matches {
+        name[..name.len() - 1].to_string()
+    } else {
+        name.clone()
+    }
+}
+
+pub fn to_regex(keywords: &Vec<String>) -> Vec<Regex> {
```
Comment (dansteren) [inline src/keyword.rs:17]: `to_regex` to me indicates that I will get one regex out. Consider changing this to `to_regex_list` (or something), or doing the mapping outside the function.

File: src/nodes/act_canister_method.rs [hunk 7]
```diff
@@ -29,11 +31,11 @@ pub fn get_all_types_from_canister_method_acts(
         })
 }
 
-impl ToTokenStream for ActCanisterMethod {
-    fn to_token_stream(&self) -> TokenStream {
+impl ToTokenStream<&Vec<String>> for ActCanisterMethod {
+    fn to_token_stream(&self, context: &Vec<String>) -> TokenStream {
```
Comment (lastmjs) [inline src/nodes/act_canister_method.rs:35]: When concretely specifying the context, use a descriptive variable name

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: src/nodes/act_external_canister/act_external_canister_method.rs [hunk 8]
```diff
@@ -11,14 +11,19 @@ pub struct ActExternalCanisterMethod {
 }
 
 impl ActExternalCanisterMethod {
-    pub fn to_token_stream(&self, canister_name: &String) -> TokenStream {
-        let call_function = self.generate_call_function(canister_name);
-        let call_with_payment_function = self.generate_call_with_payment_function(canister_name);
+    pub fn to_token_stream(
+        &self,
+        canister_name: &String,
```
Comment (lastmjs) [inline src/nodes/act_external_canister/act_external_canister_method.rs:16]: Implement the ToTokenStream trait and combine these params into one context struct

File: src/nodes/act_heartbeat_method.rs [hunk 9]
```diff
@@ -7,8 +7,8 @@ pub struct ActHeartbeatMethod {
     pub body: TokenStream,
 }
 
-impl ToTokenStream for ActHeartbeatMethod {
-    fn to_token_stream(&self) -> TokenStream {
+impl ToTokenStream<()> for ActHeartbeatMethod {
+    fn to_token_stream(&self, _context: ()) -> TokenStream {
```
Comment (lastmjs) [inline src/nodes/act_heartbeat_method.rs:11]: Just underscore

File: src/nodes/act_inspect_message_method.rs [hunk 10]
```diff
@@ -9,8 +9,8 @@ pub struct ActInspectMessageMethod {
     pub body: TokenStream,
 }
 
-impl ToTokenStream for ActInspectMessageMethod {
-    fn to_token_stream(&self) -> TokenStream {
+impl ToTokenStream<()> for ActInspectMessageMethod {
+    fn to_token_stream(&self, _context: ()) -> TokenStream {
```
Comment (lastmjs) [inline src/nodes/act_inspect_message_method.rs:13]: Just underscore

File: src/nodes/act_pre_upgrade_method.rs [hunk 11]
```diff
@@ -8,8 +8,8 @@ pub struct ActPreUpgradeMethod {
     pub body: TokenStream,
 }
 
-impl ToTokenStream for ActPreUpgradeMethod {
-    fn to_token_stream(&self) -> TokenStream {
+impl ToTokenStream<()> for ActPreUpgradeMethod {
+    fn to_token_stream(&self, _context: ()) -> TokenStream {
```
Comment (lastmjs) [inline src/nodes/act_pre_upgrade_method.rs:12]: Just underscore

File: src/nodes/data_type_nodes/act_funcs.rs [hunk 12]
```diff
@@ -57,17 +57,15 @@ impl HasMembers for ActFunc {
     }
 }
 
-impl ToTokenStream for FuncLiteral {
-    fn to_token_stream(&self) -> TokenStream {
+impl<C> ToTokenStream<C> for FuncLiteral {
+    fn to_token_stream(&self, _context: C) -> TokenStream {
```
Comment (lastmjs) [inline src/nodes/data_type_nodes/act_funcs.rs:61]: Underscore only

File: src/nodes/data_type_nodes/act_funcs.rs [hunk 13]
```diff
@@ -57,17 +57,15 @@ impl HasMembers for ActFunc {
...
         self.func.name.to_identifier().to_token_stream()
     }
 }
 
-impl ToTokenStream for FuncTypeAlias {
-    fn to_token_stream(&self) -> TokenStream {
-        generate_func_struct_and_impls(
-            &self.func,
-        )
+impl ToTokenStream<&Vec<String>> for FuncTypeAlias {
```
Comment (lastmjs) [inline src/nodes/data_type_nodes/act_funcs.rs:66]: Double-check that we aren't passing keyword_lists in where they aren't needed

File: src/nodes/data_type_nodes/act_primitives.rs [hunk 14]
```diff
@@ -52,16 +52,16 @@ pub struct ActPrimitiveTypeAlias {
...
+        let name = self.name.to_identifier();
+        let alias = self.aliased_type.to_token_stream(context);
         quote!(type #name = #alias;)
     }
 }
 
-impl ToTokenStream for ActPrimitiveLit {
-    fn to_token_stream(&self) -> TokenStream {
+impl<C> ToTokenStream<C> for ActPrimitiveLit {
+    fn to_token_stream(&self, _context: C) -> TokenStream {
```
Comment (lastmjs) [inline src/nodes/data_type_nodes/act_primitives.rs:64]: Underscore only

**Review by lastmjs — CHANGES_REQUESTED** [review]

**Review by lastmjs — APPROVED** [review]

**Pull request merged**